x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/action/rest/RestXSearchAction.java

```diff
@@ -0,0 +1,105 @@
+import org.elasticsearch.xpack.relevancesearch.xsearch.action.XSearchAction;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+import static org.elasticsearch.rest.RestRequest.Method.GET;
+import static org.elasticsearch.rest.RestRequest.Method.POST;
+
+public class RestXSearchAction extends BaseRestHandler {
+
+    public static final String REST_BASE_PATH = "/{index}/_xsearch";
+
+    private final RelevanceMatchQueryRewriter relevanceMatchQueryRewriter;
+    private final XSearchRequestValidationService xSearchRequestValidationService;
+
+    @Inject
+    public RestXSearchAction(
+        RelevanceMatchQueryRewriter relevanceMatchQueryRewriter,
+        XSearchRequestValidationService xSearchRequestValidationService
+    ) {
+        super();
+        this.relevanceMatchQueryRewriter = relevanceMatchQueryRewriter;
+        this.xSearchRequestValidationService = xSearchRequestValidationService;
+    }
+    @Override
+    public List<Route> routes() {
+        return List.of(new Route(GET, REST_BASE_PATH), new Route(POST, REST_BASE_PATH));
+    }
+
+    @Override
+    public String getName() {
+        return "xsearch_action";
+    }
+
+    @Override
+    public RestChannelConsumer prepareRequest(final RestRequest request, final NodeClient client) throws IOException {
+        String index = request.param("index");
+        boolean explain = request.paramAsBoolean("explain", false);
+        XContentParser parser = request.contentOrSourceParamParser();
+        XSearchAction.Request xsearchRequest = XSearchAction.Request.parseRequest(index, parser, explain);
+        xsearchRequest.indicesOptions(IndicesOptions.fromRequest(request, xsearchRequest.indicesOptions()));
+
+        xSearchRequestValidationService.validateRequest(xsearchRequest);
+
+        RelevanceMatchQueryBuilder queryBuilder = new RelevanceMatchQueryBuilder(relevanceMatchQueryRewriter, xsearchRequest.getQuery());
+        return channel -> doXSearch(index, xsearchRequest.explain(), queryBuilder, client, channel);
+    }
+
+    private static void doXSearch(
+        String index,
+        boolean explain,
+        RelevanceMatchQueryBuilder queryBuilder,
+        NodeClient client,
+        RestChannel channel
+    ) {
+        SearchRequest searchRequest = client.prepareSearch(index)
+            .setQuery(queryBuilder)
+            .setSize(1000)
+            .setFetchSource(true)
+            .setExplain(explain)
+            .request();
+
+        client.execute(SearchAction.INSTANCE, searchRequest, new RestBuilderListener<>(channel) {
+            @Override
+            public RestResponse buildResponse(SearchResponse searchResponse, XContentBuilder builder) throws Exception {
+                searchResponse.toXContent(builder, ToXContent.EMPTY_PARAMS);
+                return new RestResponse(RestStatus.OK, builder);
+            }
+        });
```

Comment:
I think we need to invoke the `XSearchAction.INSTANCE` here, right? We intend to execute the `XSearchTransportAction` that in turn executes the `SearchAction`...

Also, I believe we could use `RestStatusToXContentListener` for retrieving the right response status:

```suggestion
        client.execute(XSearchAction.INSTANCE, searchRequest, new RestStatusToXContentListener<>(channel));
```

Reply:
Thanks for the suggestion, added!